source/AST/ASTVisitor.cpp

```diff
@@ -1582,58 +1461,251 @@ TraverseFunctionTemplateDecl(
     // (e.g. for an abbreviated function template with no template-head)
     if(! shouldExtract(FD))
         return true;
+    auto Template = std::make_unique<TemplateInfo>();
+    parseTemplateParams(*Template, FD);
 
-    FunctionInfo I;
-    parseTemplateParams(I.Template, FD);
+    // traverse the templated declaration according to its kind
+    return TraverseDecl(FD, std::move(Template));
+}
 
-    switch(FD->getKind())
+bool
+ASTVisitor::
+Traverse(
+    ClassScopeFunctionSpecializationDecl* D)
+{
+    if(! shouldExtract(D))
+        return true;
+
+    // for class scope explicit specializations of member function templates which
```

Comment:
I would use /* */ to allow easier editing. And I would use proper grammar and capitalization since it is so much exposition.